src/diagnostics.rs

```diff
@@ -0,0 +1,204 @@
+                    symbol.name(),
+                    symbol.lineno(),
+                    symbol.addr()
+                )));
+            });
+            true
+        });
+        diag.unwrap()
+    }
+
+    pub fn error<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Error, message.into())
+    }
+
+    pub fn warning<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Warning, message.into())
+    }
+
+    pub fn note<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Note, message.into())
+    }
+
+    pub fn help<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Help, message.into())
+    }
+}
+
+#[must_use = "call `emit` to emit the diagnostic"]
+pub struct DiagnosticBuilder<'a: 'b, 'b> {
+    file_map: &'b FileMap<&'a str>,
+    out: &'b mut StandardStream,
+    diagnostics: Vec<Diagnostic>,
+
+    severity: Severity,
+    message: String,
+    code: Option<String>,
+    labels: Vec<Label>,
+}
+
+impl<'a: 'b, 'b> DiagnosticBuilder<'a, 'b> {
+    pub fn emit(self) {
+        let Self { file_map, out, mut diagnostics, severity, message, code, labels } = self;
+        diagnostics.push(Diagnostic { severity, message, code, labels });
+
+        // ignore output errors, because where would we log them anyway?!
+        for diagnostic in diagnostics {
+            let _ = codespan_reporting::emit_single(&mut *out, file_map, &diagnostic);
+        }
+        writeln!(out);
+    }
+
+    fn diagnostic(self, new_severity: Severity, new_message: String) -> Self {
+        let Self { file_map, out, mut diagnostics, severity, message, code, labels } = self;
+        diagnostics.push(Diagnostic { severity, message, code, labels });
+
+        Self {
+            file_map,
+            out,
+            diagnostics,
+            severity: new_severity,
+            message: new_message,
+            code: None,
+            labels: Vec::new(),
+        }
+    }
+
+    pub fn bug<S: Into<String>>(self, message: S) -> Self {
+        self.diagnostic(Severity::Bug, message.into())
+    }
+
+    pub fn error<S: Into<String>>(self, message: S) -> Self {
+        self.diagnostic(Severity::Error, message.into())
+    }
+
+    pub fn warning<S: Into<String>>(self, message: S) -> Self {
+        self.diagnostic(Severity::Warning, message.into())
+    }
+
+    pub fn note<S: Into<String>>(self, message: S) -> Self {
+        self.diagnostic(Severity::Note, message.into())
+    }
+
+    pub fn help<S: Into<String>>(self, message: S) -> Self {
+        self.diagnostic(Severity::Help, message.into())
+    }
+
+    pub fn with_error_code(mut self, code: String) -> Self {
+        self.code = Some(code);
+        self
+    }
+
+    fn with_section<S: Into<String>>(
+        mut self, style: LabelStyle, range: &Range<usize>, message: S,
```

Comment:
`Range` doesn't implement `Copy`. For easier usability I accept a reference here, instead of requiring all users to always call `.with_section(range.clone(), "foo")`. I guess as discussed in https://github.com/oberien/heradoc/pull/69#issuecomment-478477151 a `SourceRange` type could make sense to get rid of this kind of stuff by implementing `Copy` on it.

Reply:
Since the only place where you'd construct them by hand and thus really benefit from language semantics (`..`) is within the tests, that would be quite reasonable. The most immediate other benefit is being able to use it as an index parameter but the final usage of `Range` currently is constructing the argument to the diagnostics library.